axum-jwks/src/key_manager.rs

```diff
@@ -0,0 +1,143 @@
+use std::sync::Arc;
+
+use crate::key_store::JwksError;
+use crate::{KeyStore, TokenError};
+use jsonwebtoken::{decode, decode_header, TokenData};
+use serde::de::DeserializeOwned;
+
+use tokio::sync::RwLock;
+
+use tokio::time::{sleep, Duration, Instant};
+use tracing::{debug, error, info};
+
+#[derive(Clone)]
+pub struct KeyManager {
+    authority: String,
+    audience: String,
+    minimal_interval: Option<Duration>,
+    key_store: Arc<RwLock<KeyStore>>,
+    client: reqwest::Client,
+}
+
+impl KeyManager {
```

Comment:
Instead of constructing an empty keystore and then waiting for the first update interval, it might make more sense to employ a builder pattern, and to fetch the keys synchronously the first time. This should make it easier to expose errors on the first fetch. I think it's much more likely that the first fetch errors (maybe a configuration issue) rather than future ones, so this should make for a better developer experience.

Reply:
Considered that, wanted simplicity and being free to set it up more freely. I now agree with you. Better to make a more standard builder pattern. Users are free to create it as they wish without using the builder pattern